File: C_Projects/Asignacion #3 (Comparacion de Lenguajes)/binary_tree.c
/*
 * binary_tree.c
 * By Antonio F. Huertas
 * Edited by Saulo Almeda Nieves
 * 841-13-0170
 * 11 / 2 /2015
 * This program defines and manipulates a binary search tree.
 */

#include <stdio.h>   // for puts, printf, scanf
#include <stdbool.h> // for bool, true, false
#include <stddef.h>  // for NULL
#include <stdlib.h>  // for malloc, free

// Defines an alias for the node structure.
typedef struct node Node;

// Declares a structure for nodes.
struct node {
    int data;
    Node *left, *right;
};

// Declares the data structure for a binary tree.
typedef struct {
    Node *root;
} Tree;

// Declares functions using prototypes.
Tree *create_tree(void);
void add_tree(Tree *bst, int val);
void show_tree(const Tree *bst);
int sum_tree(const Tree *bst);
bool search_tree(const Tree *bst, int val);
void destroy_tree(Tree *bst);

Node *add_helper(Node *curr, int val);
void show_helper(const Node *curr);
int sum_helper(const Node *curr);
bool search_helper(Node *curr, int val);
void destroy_helper(Node *curr);

//Added Function Prototypes
int size_tree(const Tree *bst);
int depth_tree(const Tree *bst);
int minimum_tree(const Tree *bst);

// Starts the execution of the program.
int main(void) {
    Tree *bst = create_tree();
    puts("A binary search tree has been created.");

    add_tree(bst, 40);
    add_tree(bst, 10);
    add_tree(bst, 20);
    add_tree(bst, 50);
    add_tree(bst, 30);

    printf("After adding some elements, the tree is ");
    show_tree(bst);
    printf("\nThe sum of its elements is %d\n", sum_tree(bst));
    printf("\n");

    int num;
    printf("Enter a number to be searched: ");
    scanf("%d", &num);

    bool found = search_tree(bst, num);
    if (found)
        puts("The number was found!");
    else
        puts("The number was NOT found.");

    printf("\nEnter a number to be added to the tree: ");
    scanf("%d", &num);
    add_tree(bst, num);

    printf("After adding a new element, the tree is ");
    show_tree(bst);
    printf("\nThe sum of its elements is %d\n", sum_tree(bst));
    printf("\n");

    printf("\nThe size of the tree is: %d\n", size_tree(bst));

    printf("\nThe minimum element of the tree is: %d\n", minimum_tree(bst));

    destroy_tree(bst);
    bst = NULL;
    puts("The binary search tree has been deallocated.");
    return 0;
}

//Returns the number of elements in the given tree.
int size_tree(const Tree *bst) {
    int size = 0;
    Node *curr_right = bst->root->right;
    Node *curr_left = bst->root->left;
    if(curr_left->left == NULL)
        return size++;
    else if(curr_right->right == NULL)
        return size++;
    else
        return size_tree(curr_left->left) + size_tree(curr_right->right);
}

//Returns the number of levels in the tree.
int depth_tree(const Tree *bst) {
    int levels = 0;
}

//Returns the minimum element in the tree.
int minimum_tree(const Tree *bst) {
    Node *curr = bst->root->left;
    while(curr->left != NULL)
        curr = curr->left;
    return (curr->data);
}

// Returns a new empty tree.
Tree *create_tree(void) {
    Tree *bst = malloc(sizeof(Tree));
    bst->root = NULL;
    return bst;
}

// Adds a new value to the given tree.
void add_tree(Tree *bst, int val) {
    bst->root = add_helper(bst->root, val);
}

Node *add_helper(Node *curr, int val) {
    if (curr == NULL) {
        curr = malloc(sizeof(Node));
        curr->data = val;
        curr->left = curr->right = NULL;
    } else {
        if (val < curr->data)
            curr->left = add_helper(curr->left, val);
        else if (val > curr->data)
            curr->right = add_helper(curr->right, val);
        else
            curr->data = val;
    }
    return curr;
}

// Shows the contents of the given tree.
void show_tree(const Tree *bst) {
    show_helper(bst->root);
}

void show_helper(const Node *curr) {
    if (curr != NULL) {
        show_helper(curr->left);
        printf("%d ", curr->data);
        show_helper(curr->right);
    }
}

// Returns the sum of the given tree.
int sum_tree(const Tree *bst) {
    return sum_helper(bst->root);
}

int sum_helper(const Node *curr) {
    if (curr == NULL)
        return 0;
    else
        return sum_helper(curr->left) +
				curr->data +
				sum_helper(curr->right);
}

// Determines whether the given tree contains the given value.
bool search_tree(const Tree *bst, int val) {
    return search_helper(bst->root, val);
}

bool search_helper(Node *curr, int val) {
    if (curr == NULL)
        return false;
    else {
        if (val < curr->data)
            return search_helper(curr->left, val);
        else if (val > curr->data)
            return search_helper(curr->right, val);
        else
            return true;
    }
}

// Frees the memory allocated for the tree.
void destroy_tree(Tree *bst) {
    destroy_helper(bst->root);
    free(bst);
}

void destroy_helper(Node *curr) {
    if (curr != NULL) {
        destroy_helper(curr->left);
        destroy_helper(curr->right);
        free(curr);
    }
}

File: F# Workspace/Asignacion 5 (Comparacion de Lenguajes)/Asignacion 5 (Comparacion de Lenguajes)/binary_tree.fs
(*
 * binary_tree.fs
 * By Antonio F. Huertas
 * Edited by Saulo Almeda Nieves
 * 841-13-0170
 * 12/2/15
 * This program defines and manipulates a binary search tree of 
 * integers.
 *)

// Defines a discriminated union type for a binary tree.   
type IntTree =
    | Nil
    | Node of int * IntTree * IntTree

// sum_tree : IntTree -> int
// Returns the sum of the elements in the given tree.
let rec sum_tree = function
    | Nil -> 0
    | Node (root, lft, rgt) -> sum_tree lft + root + sum_tree rgt

// contains_tree : IntTree -> int -> bool
// Returns true if the given tree contains the given element.    
let rec contains_tree bst elem =
    match bst with
    | Nil -> false
    | Node (root, lft, rgt) 
            when elem < root -> contains_tree lft elem
    | Node (root, lft, rgt) 
            when elem > root -> contains_tree rgt elem
    | Node (root, lft, rgt) -> true

// add_tree : IntTree -> int -> IntTree
// Returns the tree that results by adding the given element to the
// given tree.
let rec add_tree bst elem =
    match bst with
    | Nil -> Node (elem, Nil, Nil)
    | Node (root, lft, rgt) 
            when elem < root -> Node (root, (add_tree lft elem), rgt)
    | Node (root, lft, rgt) 
            when elem > root -> Node (root, lft, (add_tree rgt elem))
    | Node (root, lft, rgt) -> bst

// show_tree : IntTree -> unit
// Shows the elements in the given tree using inorder.
let rec show_tree = function
    | Nil -> ()
    | Node (root, lft, rgt) -> show_tree lft
                               printf "%d " root
                               show_tree rgt

// construct_tree : int list -> IntTree
// Returns the binary search tree that correspond to the given list.
let construct_tree lst =
    let rec helper bst lst =
        match lst with
        | [] -> bst
        | hd::[] -> Node (hd, Nil, Nil)
        | hd::tl when tl.Head < hd -> Node (hd, helper (Node (tl.Head, Nil, Nil)) tl, Nil)
        | hd::tl -> Node (hd, Nil, helper (Node (tl.Head, Nil, Nil)) tl)
    helper Nil lst

// size_tree : IntTree -> int
// Returns the number of elements in the given IntTree.
let rec size_tree = function
    | Nil -> 0
    | Node (root, Nil, Nil) -> 1
    | Node (root, lft, rgt) -> size_tree lft + size_tree rgt + 1
    
// minimum_tree : IntTree -> int
// Returns the minimum element of the given IntTree.
let minimum_tree bst =
    let rec helper min bst =
        match bst with
        | Nil -> min
        | Node (root, Nil, Nil) -> root
        | Node (root, lft, rgt) -> helper root lft 
    helper 0 bst 


// start : unit -> unit
let start () =
    let bst = Nil
    let bst1 = add_tree bst 30
    let bst2 = add_tree bst1 40
    let bst3 = add_tree bst2 10
    let bst4 = add_tree bst3 5
    let bst5 = add_tree bst4 20

    printf "This is the tree: "
    show_tree bst5
    printfn "\nThis is the tree, internally: %A" bst5
    printfn "The sum of its elements is %d" (sum_tree bst5)
    printfn "Does it contain 50? %b" (contains_tree bst5 50)
    printfn "The size of the tree is: %d" (size_tree bst5)
    printfn "The minimum element of the tree is: %d" (minimum_tree bst5)

    let lst = [154;35;63;2;5;6;7]
    printfn "Constructing tree..."
    printfn "Here is the tree: %A" (construct_tree lst)